server/jobs/moss.py

```diff
@@ -113,35 +118,26 @@ def parse_moss_results(base_url, hashed_ids, logger, pattern, template, review_t
         submissionA = Backup.query.filter_by(id=decode_id(hashidA)).one_or_none()
         submissionB = Backup.query.filter_by(id=decode_id(hashidB)).one_or_none()
 
-        # make sure this submission hasn't been processed before
-        resultB = MossResult.query.filter_by(submissionB=submissionA, submissionA=submissionB).one_or_none()
-        if resultB:
-            logger.info('Moss result #{} already exists.'.format(match))
-            continue
-
         similarityA, similarityB = parser.similarities
         rangesA, rangesB = parser.ranges[::2], parser.ranges[1::2]
         matchesA = [[int(i) for i in r.split('-')] for r in rangesA]
         matchesB = [[int(i) for i in r.split('-')] for r in rangesB]
         matchesA = recalculate_lines(submissionA, matchesA, pattern)
         matchesB = recalculate_lines(submissionB, matchesB, pattern)
-        print(matchesA)
-        print(matchesB)
         similarityA = recalculate_similarity(submissionA, matchesA, template)
         similarityB = recalculate_similarity(submissionB, matchesB, template)
-        result = MossResult.query.filter_by(submissionA=submissionA, submissionB=submissionB).one_or_none()
-        tags = []
-        if similarityA >= review_threshold or similarityB >= review_threshold:
-            tags.append('review')
-        result.similarityA = similarityA
-        result.similarityB = similarityB
-        result.matchesA = matchesA
-        result.matchesB = matchesB
-        result.tags = tags
-        # result = MossResult(submissionA=submissionA, submissionB=submissionB,
-        #     similarityA=similarityA, similarityB=similarityB,
-        #     matchesA=matchesA, matchesB=matchesB, tags=tags)
-        db.session.add(result)
+        if not hashed_ids or hashidA in hashed_ids:
+            result = MossResult(primary=submissionA, secondary=submissionB,
+                primary_matches=matchesA, secondary_matches=matchesB,
+                tags=['review'] if similarityA >= review_threshold else [], 
+                similarity=similarityA, run_time=run_time)
+            db.session.add(result)
+        if not hashed_ids or hashidB in hashed_ids:
+            result = MossResult(primary=submissionB, secondary=submissionA,
+                primary_matches=matchesB, secondary_matches=matchesA,
+                tags=['review'] if similarityB >= review_threshold else [],
+                similarity=similarityB, run_time=run_time)
```

Comment:
Instead of `MossResult(...)` why _don't_ we just do `get_or_create` minus the runtime?  If the MossResults are the same, it'll just update the runtime.  This gets rid of needing to filter out duplicates later.

Reply:
I think `get_or_create` is a Django thing. I also now think storing once for each run might make sense, so we could easily view all results from a particular run.